File: etsyCatalogs/etsyCatalogs/View/MasterViewController.m
//
//  MasterViewController.m
//  etsyCatalogs
//
//  Created by Ihonahan Buitrago on 9/1/14.
//  Copyright (c) 2014 Ihonahan Buitrago. All rights reserved.
//

#import "MasterViewController.h"
#import "DetailViewController.h"

@interface MasterViewController ()

@end

@implementation MasterViewController
            
- (void)awakeFromNib {
    [super awakeFromNib];
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad) {
        self.clearsSelectionOnViewWillAppear = NO;
        self.preferredContentSize = CGSizeMake(420.0, 600.0);
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    UIBarButtonItem *pageUpButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemRewind target:self action:@selector(previousPage)];
    self.navigationItem.leftBarButtonItem = pageUpButton;
    UIBarButtonItem *pageDownButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFastForward target:self action:@selector(nextPage)];
    self.navigationItem.rightBarButtonItem = pageDownButton;
    self.detailViewController = (DetailViewController *)[[self.splitViewController.viewControllers lastObject] topViewController];
    
    self.navigationItem.titleView.frame = self.searchText.frame;

    [self.loadingView stopAnimating];
    self.loadingView.hidden = YES;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)nextPage
{
    [self.catalogsDatasource fastForwardList];
}

- (void)previousPage
{
    [self.catalogsDatasource rewindList];
}


#pragma mark - Segues

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([[segue identifier] isEqualToString:@"showDetail"]) {
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        
        if (self.catalogsDatasource.catalogResults.count) {
            DetailViewController *controller = (DetailViewController *)[[segue destinationViewController] topViewController];
            CatalogItem *item = [self.catalogsDatasource.catalogResults getCatalogItemByIndex:indexPath.row];
            [controller setCatalogDetail:item];
            controller.navigationItem.leftBarButtonItem = self.splitViewController.displayModeButtonItem;
            controller.navigationItem.leftItemsSupplementBackButton = YES;
        }
        
    }
}


#pragma mark - UITextFieldDelegate methods
- (BOOL)textFieldShouldBeginEditing:(UITextField *)textField
{
    return YES;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    NSString *keywords = [textField.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    if (![keywords isEqualToString:@""]) {
        [self.catalogsDatasource getInitialListForKeywords:keywords];
        self.loadingView.hidden = NO;
        [self.loadingView startAnimating];
    }
    
    [textField resignFirstResponder];
    
    return YES;
}



#pragma mark - CatalogDataSourceDelegate methods
- (void)catalogsDatasource:(CatalogsDatasource *)sender DidFailWithError:(NSError *)error
{
    
}

- (void)catalogsDatasource:(CatalogsDatasource *)sender DidLoadListingCatalogs:(int)count
{
    if (count) {
        [self.loadingView stopAnimating];
        self.loadingView.hidden = YES;
    }
}


@end

File: etsyCatalogsSwift/etsyCatalogsSwift/Controller/MasterViewController.swift
//
//  MasterViewController.swift
//  etsyCatalogsSwift
//
//  Created by Ihonahan Buitrago on 9/19/14.
//  Copyright (c) 2014 Ihonahan Buitrago. All rights reserved.
//

import UIKit

class MasterViewController: UITableViewController, CatalogsDataSourceDelegate, UITextFieldDelegate {

    @IBOutlet weak var searchText : UITextField!
    @IBOutlet weak var catalogsDatasource : CatalogsDataSource!
    @IBOutlet weak var loadingView : UIActivityIndicatorView!


    var detailViewController: DetailViewController? = nil
    var objects = NSMutableArray()


    override func awakeFromNib() {
        super.awakeFromNib()
        if UIDevice.currentDevice().userInterfaceIdiom == .Pad {
            self.clearsSelectionOnViewWillAppear = false
            self.preferredContentSize = CGSize(width: 420.0, height: 600.0)
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        let pageUpButton : UIBarButtonItem = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.Rewind,
            target: self,
            action: Selector(previousPage()))
        self.navigationItem.leftBarButtonItem = pageUpButton
        
        let pageDownButton : UIBarButtonItem = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.FastForward,
            target: self,
            action: Selector(nextPage()))
        self.navigationItem.rightBarButtonItem = pageDownButton
        
        self.navigationItem.titleView?.frame = self.searchText.frame
        
        self.loadingView.stopAnimating()
        self.loadingView.hidden = true
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    
    // MARK: - Private methods
    private func previousPage() {
        self.catalogsDatasource.rewindPage()
        self.loadingView.startAnimating()
        self.loadingView.hidden = false
    }
    
    private func nextPage() {
        self.catalogsDatasource.fastForwardPage()
        self.loadingView.startAnimating()
        self.loadingView.hidden = false
    }


    // MARK: - Segues

    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "ShowDetail" {
            if self.catalogsDatasource.catalogResults.count() > 0 {
                let detailButton = sender as UIButton
                if let item : CatalogItem = self.catalogsDatasource.catalogResults.getCatalogItemByIndex(detailButton.tag)? {
                    let detailViewController : DetailViewController = segue.destinationViewController.topViewController as DetailViewController
                    detailViewController.detailItem = item
                    detailViewController.navigationItem.leftBarButtonItem = self.splitViewController?.displayModeButtonItem()
                    detailViewController.navigationItem.leftItemsSupplementBackButton = true
                }
                
                
            }
        }
    }
    
    
    // MARK: - UITextFieldDelegate methods
    func textFieldShouldBeginEditing(textField: UITextField) -> Bool {
        return true
    }
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        let keywords = textField.text.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceAndNewlineCharacterSet())
        if keywords != "" {
            self.catalogsDatasource.getInitialListForKeywords(keywords)
            self.loadingView.startAnimating()
            self.loadingView.hidden = false
        }
        
        textField.resignFirstResponder()
        
        return true
    }
    
    
    // MARK: - CatalogsDataSourceDelegate methods
    func catalogsDataSourceDidFailWithError(catalogDataSource: CatalogsDataSource, error: NSError?) {
        
    }

    func catalogsDataSourceDidLoadListingCatalogs(catalogDataSource: CatalogsDataSource, count: Int) {

    }
    
    func catalogsDataSourceSelectedDetailsForListingImage(catalogDataSource: CatalogsDataSource, catalogItem: CatalogItem) {

    }

}

